### ISSUE TITLE
Unsoundness due to invalid inheritance being permitted

### ISSUE THREAD

**Issue by abgruszecki** [lampepfl/dotty#5817]
```scala
object Test {
  trait A[+X] { def x: X }
  class B[+X](val x: X) extends A[X]
  class C[+X](x: Any) extends B[Any](x) with A[X]

  def get(a: A[Int]): Int = a.x
  get(new C[Int]("foo")) // class cast exception: String is not Int
}
```

Note that we previously came to the conclusion that we do not want to prevent inheritance like `C` in general: #3989.

The likely fix would be to require `C` to override `B#x` - this is precisely what Scalac does.

**Comment by smarter** [comment]
The fix for #3989 was done in refchecks so it would make sense to do the same here I guess.

**Comment by odersky** [comment]
Rename the `x` to `y` in `C` and it stops compiling:
```scala
object Test {
  trait A[+X] { def x: X }
  class B[+X](val x: X) extends A[X]
  class C[+X](y: Any) extends B[Any](y) with A[X]

  def get(a: A[Int]): Int = a.x
  get(new C[Int]("foo")) // class cast exception: String is not Int
}
```
You get:
```
4 |  class C[+X](y: Any) extends B[Any](y) with A[X]
  |        ^
  |        value x in class B is not a legal implementation of `x' in class C
  |          its type             Any
  |          does not conform to  => X
```
There's a special code path in PostTyper that merges the values of `x` in `B` and `C` in order to avoid creating unnecessary field slots. That logic seems to be missing a test.

**Comment by smarter** [comment]
> There's a special code path in PostTyper that merges the values of x in B and C in order to avoid creating unnecessary field slots. 

Link for anyone looking for it: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/ParamForwarding.scala

**Comment by dwijnand** [comment]
Looks like this is fixed in at least 3.0.1:

```
$ scalac -3 i5817.scala
https://repo1.maven.org/maven2/io/get-coursier/apps/maven-metadata.xml
  No new update since 2021-09-02 12:27:05
-- [E163] Declaration Error: i5817.scala:4:8 -----------------------------------
4 |  class C[+X](x: Any) extends B[Any](x) with A[X]
  |        ^
  |        [error overriding method ](url)x in trait A of type => X;
  |          value x in class B of type Any has incompatible type

longer explanation available when compiling with `-explain`
1 error found
```

Maybe with #11133?
